src/platform/libmach.zig

```diff
@@ -2,34 +2,106 @@ const std = @import("std");
 const Core = @import("../Core.zig");
 const gpu = @import("gpu");
 const ecs = @import("ecs");
+const glfw = @import("glfw");
 
 pub const App = @This();
 
-// Zig says that *App has a size of 0 bits, and it won't compile if
-// pub const core_callback_t = fn (*App, *Core) callconv(.C) void;
-// What is *App needed for anyway?
-pub const CoreCallback = fn (*Core) callconv(.C) void;
-
-pub const CoreCallbacks = struct {
-    core_init: ?CoreCallback,
-    core_update: ?CoreCallback,
-    core_deinit: ?CoreCallback,
-};
-
-pub var core_callbacks = CoreCallbacks {
-    .core_init = null,
-    .core_update = null,
-    .core_deinit = null,
-};
-
-pub fn init(_: *App, core: *Core) !void {
-    core_callbacks.core_init.?(core);
+// Dummy init, deinit, and update functions
+pub fn init(_: *App, _: *Core) !void { }
+
+pub fn deinit(_: *App, _: *Core) void { }
+
+pub fn update(_: *App, _: *Core) !void { }
+
+// Current Limitations:
+// 1. Currently, ecs seems to be using some weird compile-time type trickery, so I'm not exactly sure how
+// `engine` should be integrated into the C API
+// 2. Core might need to expose more state so more API functions can be exposed (for example, the WebGPU API)
+// 3. Be very careful about arguments, types, memory, etc - any mismatch will result in undefined behavior
+
+pub export fn mach_set_should_close(core: *Core) void {
+    core.setShouldClose(true);
+}
+
+pub export fn mach_delta_time(core: *Core) f32 {
+    return core.delta_time;
+}
+
+var gpa = std.heap.GeneralPurposeAllocator(.{}){};
+const allocator = gpa.allocator();
+
+// Initializes and returns Mach's core structure
+// Uses an optional type so it is valid to return 0 (on an error)
+// TODO: come up with a better error reporting system
+pub export fn mach_init_core() ?*Core {
+    const core: *Core = allocator.create(Core) catch {
+        return @intToPtr(?*Core, 0); // on error, return null pointer
+    };
+    core.* = Core.init(allocator) catch {
+        return @intToPtr(?*Core, 0); // on error, return null pointer
+    };
+
+    // // Glfw specific: initialize the user pointer used in callbacks
+    core.*.internal.initCallback();
+
+    return core;
 }
 
-pub fn deinit(_: *App, core: *Core) void {
-    core_callbacks.core_deinit.?(core);
+// Deinitializes mach core structure
+pub export fn mach_deinit(core: *Core) void {
+    core.internal.deinit();
+    allocator.destroy(core);
 }
 
-pub fn update(_: *App, core: *Core) !void {
-    core_callbacks.core_update.?(core);
+pub export fn mach_window_should_close(core: *Core) bool {
+    return core.internal.window.shouldClose();
+}
+
+pub const CoreCallback = fn (*Core, u32, u32) callconv(.C) void;
+
+// Adapted from native.zig
```

Comment:
I think to increase the odds that `libmach` is always up-to-date and working, and to get higher odds of it also working on other platforms (webassembly for example), we should instead have `native.zig` export an API which `libmach.zig` calls into. Then `libmach.zig` becomes just rather simple C export wrappers.

@iddev5 do you agree with that?

Reply:
Yes, this was the idea all along. Plus these functions will be able to reduce the code duplication which we have with mach_update and things like that.